File: 701-800/728-self-dividing-numbers.py
class Solution:
    def selfDividingNumbers(self, left: int, right: int) -> List[int]:
        is_self_dividing = lambda num: all(digit != '0' and num % int(digit) == 0 for digit in str(num))
        return filter(is_self_dividing, range(left, right + 1))

File: 701-800/728-self-dividing-numbers.java
// O(nm)
// n = number between left to right
// m = digit of each number
class Solution {
  public List<Integer> selfDividingNumbers(int left, int right) {
      List<Integer> result = new ArrayList<Integer>();
      for(int i=left;i<=right;i++) {
          if(isSelfDividing(i)) {
              result.add(i);
          }
      }
      return result;
  }
  
  public boolean isSelfDividing(int num) {
      int number = num;
      while(number > 0) {
          int digit = number % 10;
          
          // self dividing num cannot contain 0
          if(digit == 0 || num % digit != 0) {
              return false;
          }
          
          number /= 10;
      }
      
      return true;
  }
}